src/tabmat/formula.py

```diff
@@ -0,0 +1,582 @@
+        self.name = name
+        return self
+
+
+class _InteractableCategoricalVector(_InteractableVector):
+    def __init__(
+        self,
+        codes: numpy.ndarray,
+        categories: List[str],
+        multipliers: numpy.ndarray,
+        name: Optional[str] = None,
+    ):
+        # sentinel values for codes:
+        # -1: missing
+        # -2: drop
+        self.codes = codes
+        self.categories = categories
+        self.multipliers = multipliers
+        self.name = name
+
+    @classmethod
+    def from_categorical(
+        cls, cat: pandas.Categorical, reduced_rank: bool
+    ) -> "_InteractableCategoricalVector":
+        """Create an interactable categorical vector from a pandas categorical."""
+        categories = list(cat.categories)
+        codes = cat.codes.copy().astype(numpy.int64)
+        if reduced_rank:
+            codes[codes == 0] = -2
+            codes[codes > 0] -= 1
+            categories = categories[1:]
+        return cls(
+            codes=codes,
+            categories=categories,
+            multipliers=numpy.ones(len(cat.codes)),
+        )
+
+    def __rmul__(self, other):
+        if isinstance(other, (int, float)):
+            return _InteractableCategoricalVector(
+                categories=self.categories,
+                codes=self.codes,
+                multipliers=self.multipliers * other,
+                name=self.name,
+            )
+
+    def to_tabmat(
+        self,
+        dtype: numpy.dtype = numpy.float64,
+        sparse_threshold: float = 0.1,
+        cat_threshold: int = 4,
+    ) -> Union[CategoricalMatrix, SplitMatrix]:
+        codes = self.codes.copy()
+        categories = self.categories.copy()
+        if -2 in self.codes:
+            codes[codes >= 0] += 1
+            codes[codes == -2] = 0
+            categories.insert(0, "__drop__")
+            drop_first = True
+        else:
+            drop_first = False
+
+        cat = pandas.Categorical.from_codes(
+            codes=codes,
+            categories=categories,
+            ordered=False,
+        )
+
+        categorical_part = CategoricalMatrix(cat, drop_first=drop_first, dtype=dtype)
+
+        if (self.codes == -2).all():
+            # All values are dropped
+            return DenseMatrix(numpy.empty((len(codes), 0), dtype=dtype))
+        elif (self.multipliers == 1).all() and len(categories) >= cat_threshold:
+            return categorical_part
+        else:
+            sparse_matrix = sps.csc_matrix(
+                categorical_part.tocsr().multiply(self.multipliers[:, numpy.newaxis])
+            )
+            (
+                dense_part,
+                sparse_part,
+                dense_idx,
+                sparse_idx,
+            ) = _split_sparse_and_dense_parts(sparse_matrix, sparse_threshold)
+            return SplitMatrix([dense_part, sparse_part], [dense_idx, sparse_idx])
+
+    def get_names(self) -> List[str]:
```

Comment:
I am a bit confused by the naming convention with `name` and `names`. I see how these two concepts are necessary, but I think that a better naming of `names` can be done :). What about `levels`?

Reply:
Yeah, I see your point :) The issue is that the same method name must be used for all `InteractableVector` subclasses, and `InteractableSparseVector` and `InteractableDenseVector` do not really have levels. Maybe `get_column_names`? Or even `get_feature_names`? Or are those that still too ambiguous?